generate_precompile.jl

```diff
@@ -0,0 +1,57 @@
+using SnoopCompile
+
+@snoopiBot "BinaryBuilder" begin
+    using BinaryBuilder
+    # Do an actual build
+    products = Product[
+        ExecutableProduct("hello_world_c", :hello_world_c),
+        ExecutableProduct("hello_world_cxx", :hello_world_cxx),
+        ExecutableProduct("hello_world_fortran", :hello_world_fortran),
+        ExecutableProduct("hello_world_go", :hello_world_go),
+        ExecutableProduct("hello_world_rust", :hello_world_rust),
+    ]
+
+    # First, do the build, but only output the meta json, since we definitely want that to be fast
+    build_tarballs(
+        ["--meta-json=/dev/null"],
+        "testsuite",
+        v"1.0.0",
+        # No sources
+        DirectorySource[],
```

Comment:
I don't know much about SnoopCompile, but does it make sense to use a different type of source here since DirectorySource is already used above?

Reply:
I don't think the compilation time of `DirectorySource()` really matters here.... but I could be wrong!

Reply:
I was more thinking about the signature of the method of `build_tarballs` being called, I thought that maybe varying the signature would precompile more methods

Reply:
Perhaps we should instead `@nospecialize` things, since this really isn't something we want to fight with that much.